lib/src/test/java/org/kpipe/consumer/OffsetManagerTest.java

```diff
@@ -0,0 +1,798 @@
+import java.util.Map;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.TopicPartition;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.kpipe.consumer.enums.ConsumerCommand;
+import org.mockito.*;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+class OffsetManagerTest {
+
+  private static final String TOPIC = "test-topic";
+
+  @Mock
+  private Consumer<String, String> mockConsumer;
+
+  @Captor
+  private ArgumentCaptor<Map<TopicPartition, OffsetAndMetadata>> offsetCaptor;
+
+  private OffsetManager<String, String> offsetManager;
+  private BlockingQueue<ConsumerCommand> commandQueue;
+  private final TopicPartition partition = new TopicPartition(TOPIC, 0);
+
+  @BeforeEach
+  void setup() {
+    commandQueue = new LinkedBlockingQueue<>();
+    offsetManager = OffsetManager.builder(mockConsumer).withCommandQueue(commandQueue).build();
+    offsetManager.start();
+  }
+
+  @AfterEach
+  void tearDown() {
+    if (offsetManager != null) {
+      try {
+        offsetManager.close();
+      } catch (final Exception ignored) {}
+      offsetManager = null;
+    }
+
+    if (commandQueue != null) {
+      commandQueue.clear();
+      commandQueue = null;
+    }
+  }
+
+  @Test
+  void shouldHandleEmptyPartitions() throws Exception {
+    final var result = offsetManager.commitSyncAndWait(1);
+
+    assertTrue(result, "Should handle empty partitions gracefully");
+    verify(mockConsumer, never()).commitSync(anyMap());
+  }
+
+  @Nested
+  class OffsetCommitTests {
+
+    @ParameterizedTest
+    @MethodSource("org.kpipe.consumer.OffsetManagerTest#contiguousOffsetScenarios")
+    void shouldOnlyCommitContiguousOffsetsWhenProcessedOutOfOrder(
+      final String description,
+      final long initialOffset,
+      final List<Long> offsetsToTrack,
+      final List<Long> offsetsToProcess,
+      final long expectedCommittedOffset
+    ) throws Exception {
+      // Track the initial offset
+      final var record = new ConsumerRecord<>(TOPIC, 0, initialOffset, "key", "value");
+      offsetManager.trackOffset(record);
+
+      // Process in batches
+      offsetsToTrack.forEach(offset -> offsetManager.trackOffset(partition, offset));
+      offsetsToProcess.forEach(offset -> offsetManager.markOffsetProcessed(partition, offset));
```

Comment:
this markOffsetProcessed overload is not doing +1 - while the other overload - that is called by actual FunctionalConsumer code does - which makes the test not representative of code execution in the application...
so asserts are basically off by 1.

Reply:
similarly other tests that use those overloads are off...

Reply:
Fixed